### PR TITLE
Expose OOM event count to prometheus

### LINKED ISSUE #1837: Getting docker container oom event.
Hello. 

Is there a way to get container oom event as a prometheus metric?
I am using cAdvisor to monitor docker containers with Prometheus and Grafana. I can see that there is a **oom_kill_event** in the API, but cannot find something similar in prometheus metrics.

### PR DESCRIPTION
**Pull request by kragniz**
Fixes #1837.

### CODE REVIEW

**Comment by kragniz** [comment]
I'm not convinced this is the best way to get the event data through, but seemed like the most straightforward. If this is an acceptable method, I can add tests and tidy this up further.

File: manager/manager.go [hunk 1]
```diff
@@ -1237,6 +1237,24 @@ func (m *manager) watchForNewOoms() error {
...
+			if err != nil {
+				klog.Errorf("failed getting container info for %q: %v", oomInstance.ContainerName, err)
+				continue
+			}
+			if len(conts) != 1 {
+				klog.Errorf("Expected the request to match only one container")
+				continue
+			}
+			for _, cont := range conts {
+				cont.oomEvents++
```
Comment (iwankgb) [inline manager/manager.go:1256]: [It is not safe](https://golang.org/ref/mem) to increment this counter from one goroutine while another might read from it.

File: manager/manager.go [hunk 2]
```diff
@@ -1237,6 +1237,24 @@ func (m *manager) watchForNewOoms() error {
...
 			}
+
+			// Count OOM events for later collection by prometheus
+			request := v2.RequestOptions{
+				IdType: v2.TypeName,
+				Count:  1,
+			}
+			conts, err := m.getRequestedContainers(oomInstance.ContainerName, request)
+			if err != nil {
+				klog.Errorf("failed getting container info for %q: %v", oomInstance.ContainerName, err)
```
Comment (iwankgb) [inline manager/manager.go:1248]: This is not an error if it's fine to `continue`.

File: manager/manager.go [hunk 3]
```diff
@@ -1237,6 +1237,24 @@ func (m *manager) watchForNewOoms() error {
...
+				continue
+			}
+			if len(conts) != 1 {
+				klog.Errorf("Expected the request to match only one container")
```
Comment (iwankgb) [inline manager/manager.go:1252]: Same here - this is not an error.

File: manager/manager.go (hunk 1 above)
Comment (kragniz) [reply manager/manager.go:1256]: What change do you suggest? Replace the increment with `atomic.AddUint64`?

File: manager/manager.go (hunk 2 above)
Comment (kragniz) [reply manager/manager.go:1248]: Sure, what's the wanted log level? (`Warningf`?)

File: cmd/cadvisor.go [hunk 4]
```diff
@@ -112,6 +112,7 @@ var (
 		container.CPUTopologyMetrics:             struct{}{},
 		container.ResctrlMetrics:                 struct{}{},
 		container.CPUSetMetrics:                  struct{}{},
+		container.OomMetrics:                     struct{}{},
```
[the remark this answers is not in the document]
Comment (iwankgb) [reply cmd/cadvisor.go:115]: We can't change names of existing exported fields, but we should not add new that repeat old errors :)

File: manager/manager.go (hunk 2 above)
Comment (iwankgb) [reply manager/manager.go:1248]: `klog.V(2).Info()` could make some sense, I think.

File: manager/manager.go (hunk 1 above)
Comment (iwankgb) [reply manager/manager.go:1256]: Or using a lock. Keep in mind that no matter what you do you will have to serialize reads too.

Comment (kragniz) [reply manager/manager.go]: Moved it to atomics

**Comment by kragniz** [comment]
@iwankgb thanks for the reviews, I've updated with fixes

**Review by iwankgb — APPROVED** [review]

**Comment by darkl0rd** [comment]
This metric (Docker) always returns 0 for me - see also #2622.

**Pull request merged**